src/writer/defaultIndexData.h

```diff
@@ -21,78 +21,100 @@
 #include "xapian/myhtmlparse.h"
 #include "tools.h"
 
+#include <atomic>
+#include <mutex>
+#include <sstream>
+
 namespace zim
 {
   namespace writer
   {
     class DefaultIndexData : public IndexData {
       public:
-        DefaultIndexData(const std::string& htmlData, const std::string& title)
-          : title(title)
-        {
+        DefaultIndexData(std::unique_ptr<ContentProvider> contentProvider, const std::string& title)
+          : m_initialized(false),
+            mp_contentProvider(std::move(contentProvider)),
+#if defined(ENABLE_XAPIAN)
+            m_title(zim::removeAccents(title)),
+#else
+            m_title(""),
+#endif
+            m_hasIndexData(false),
+            m_content(""),
+            m_keywords(""),
+            m_wordCount(0),
+            m_geoPosition(std::make_tuple(false, 0, 0))
+        {}
+
+        void initialize() const {
+          if (m_initialized) {
+            return;
+          }
+          std::lock_guard<std::mutex> lock(m_initLock);
 #if defined(ENABLE_XAPIAN)
           try {
-            htmlParser.parse_html(htmlData, "UTF-8", true);
+            std::ostringstream ss;
+            while (true) {
+              auto blob = mp_contentProvider->feed();
+              if(blob.size() == 0) {
+                break;
+              }
+              ss << blob;
+            }
+            MyHtmlParser htmlParser;
+            htmlParser.parse_html(ss.str(), "UTF-8", true);
+            m_hasIndexData = (htmlParser.dump.find("NOINDEX") == std::string::npos);
+            m_content = zim::removeAccents(htmlParser.dump);
+            m_keywords = zim::removeAccents(htmlParser.keywords);
+            m_wordCount = countWords(htmlParser.dump);
+            if(htmlParser.has_geoPosition) {
+              m_geoPosition = std::make_tuple(true, htmlParser.latitude, htmlParser.longitude);
+            }
           } catch(...) {}
 #endif
+          m_initialized = true;
         }
```

Comment:
If `initialize()` is called on an un-initialized object from two or more threads the body of this function will be executed twice (or more). Though the executions will be serialized, the second and subsequent executions will use an empty content provider and will reset the index data to an empty state. `m_initialized` must be checked under the lock (too).